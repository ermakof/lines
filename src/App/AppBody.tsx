import React, { FC } from 'react';
import styled from '@emotion/styled';
import GameField from '@src/modules/GameField';
import { useSelector } from 'react-redux';
import HitParade from '@src/modules/HitParade';
import UserInfo from '@src/modules/UserInfo';
import { getAuth } from '@src/modules/Auth/authSaga';
import { IAuthState } from '@src/modules/Auth/model/IAuthState';

interface IRoot {
  active?: boolean;
}
const Root = styled.div<IRoot>`
  display: flex;
  height: 80vh;
  position: relative;
  overflow: auto;
  opacity: ${({ active }) => (active ? 1 : 0.3)};
`;

const AppBody: FC = () => {
  const auth: IAuthState = useSelector(getAuth);
  return (
    <Root role="bodyPanel" active={!!auth.userProfile}>
      <HitParade />
      <GameField />
      <UserInfo login={auth.userProfile?.login} />
    </Root>
  );
};

export default AppBody;
