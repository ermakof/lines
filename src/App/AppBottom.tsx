import React, { FC, memo } from 'react';
import { useSelector } from 'react-redux';

import Panel from '@src/layout/Panel';
import styled from '@emotion/styled';
import { TRootState } from '@src/store';
import Percentage from '@src/components/Percentage';
import Score from '@src/components/Score';

interface IRoot {
  active: boolean;
}
const Root = styled.div<IRoot>`
  opacity: ${({ active }) => (active ? 1 : 0.3)};
`;

const AppBottom: FC = () => {
  const { gameFieldData, score = 0 } = useSelector(({ app }: TRootState) => app);
  const isAuthorised = useSelector(({ auth }: TRootState) => !!auth.userProfile);

  return (
    <Root active={isAuthorised}>
      <Panel role="bottomPanel">
        <Score value={score} />
        <Percentage data={gameFieldData} />
      </Panel>
    </Root>
  );
};

export default memo(AppBottom);
