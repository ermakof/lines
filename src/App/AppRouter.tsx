import React from 'react';
import { useSelector } from 'react-redux';
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';

import AuthForm from '@src/modules/Auth/AuthForm';
import UserForm from '@src/modules/UserForm';
import AuthLayout from '@src/modules/Auth/AuthLayout';
import RequireAuth from '@src/modules/Auth/RequireAuth';
import { TRootState } from '@src/store';

const AppRouter = () => {
  const { userProfile } = useSelector(({ auth }: TRootState) => auth);

  return (
    <Router>
      <Routes>
        <Route element={<AuthLayout login={userProfile?.login} />}>
          <Route path="/auth" element={<AuthForm />} />
          <Route
            path="/"
            element={
              <RequireAuth login={userProfile?.login}>
                <UserForm />
              </RequireAuth>
            }
          />
        </Route>
      </Routes>
    </Router>
  );
};

export default AppRouter;
