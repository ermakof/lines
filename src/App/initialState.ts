import { IAppState } from '@src/App/model/IAppState';
import createGameField from '@src/utils/createGameField';

export const GAME_FIELD_SIZE = 9;

export const INIT_GAME_LEVEL = '1';
export const INIT_GAME_FIELD_PERCENT_FILLED = 0;

interface ISettings {
  percent: number;
  chain: number;
  increase: number;
}
interface IProps {
  [key: string]: ISettings;
}
export const GAME_LEVEL_SETTINGS: IProps = {
  '1': { percent: 10, chain: 2, increase: 3 },
  '2': { percent: 20, chain: 3, increase: 4 },
  '3': { percent: 30, chain: 4, increase: 5 },
};

const initialState: IAppState = {
  userLevel: INIT_GAME_LEVEL,
  gameFieldPercentFilled: 0,
  gameFieldData: createGameField(GAME_FIELD_SIZE ** 2, INIT_GAME_FIELD_PERCENT_FILLED),
  highlightedCells: undefined,
  score: 0,
  selectedCell: undefined,
  isLoading: false,
  hitParade: undefined,
};

export default initialState;
