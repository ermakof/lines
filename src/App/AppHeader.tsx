import React, { FC, memo } from 'react';
import Panel from '@src/layout/Panel';
import AppRouter from '@src/App/AppRouter';
import styled from '@emotion/styled';

const Root = styled(Panel)`
  width: 100%;
`;

const AppHeader: FC = () => (
  <Root role="topPanel">
    <AppRouter />
  </Root>
);

export default memo(AppHeader);
