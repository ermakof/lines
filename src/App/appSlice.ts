import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import initialState, {
  GAME_LEVEL_SETTINGS,
  INIT_GAME_FIELD_PERCENT_FILLED,
  GAME_FIELD_SIZE,
} from '@src/App/initialState';
import { createGameField } from '@src/utils/';

const appSlice = createSlice({
  name: 'app',
  initialState: initialState,
  reducers: {
    rehydrate: (state) => state,
    restoreGame: (state, action) => action.payload,
    setUserLevel: (state, { payload }: PayloadAction<string>) => {
      state.userLevel = payload;
      state.selectedCell = undefined;
      state.highlightedCells = undefined;
      state.score = 0;
      state.gameFieldPercentFilled = GAME_LEVEL_SETTINGS[payload].percent;
      state.gameFieldData = createGameField(GAME_FIELD_SIZE ** 2, state.gameFieldPercentFilled);
    },
    setSelectedCell: (state, { payload }: PayloadAction<number | undefined>) => {
      state.selectedCell = payload;
    },
    startGameSteps: (state, { payload }: PayloadAction<number>) => state,
    updateGame: (state, action) => ({ ...state, ...action.payload }),
    initApp: (state) => {
      state.gameFieldPercentFilled = INIT_GAME_FIELD_PERCENT_FILLED;
      state.gameFieldData = createGameField(GAME_FIELD_SIZE ** 2, 0);
    },
    resetApp: (state) => {
      const gameFieldPercentFilled = Number(state.userLevel) * 10;
      state.selectedCell = undefined;
      state.highlightedCells = undefined;
      state.score = 0;
      state.gameFieldPercentFilled = gameFieldPercentFilled;
      state.gameFieldData = createGameField(GAME_FIELD_SIZE ** 2, gameFieldPercentFilled);
    },
    waitOn: (state) => {
      state.isLoading = true;
    },
    waitOff: (state) => {
      state.isLoading = false;
    },
  },
});

const { actions, reducer } = appSlice;

export { appSlice, actions, reducer };
