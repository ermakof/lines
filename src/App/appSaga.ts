import { call, put, takeEvery, select } from 'redux-saga/effects';
import { actions as appActions, actions } from '@src/App/appSlice';
import { LOCAL_STORAGE_APP_KEY } from '@src/store';

function* watchRehydrate(): Generator {
  yield put(appActions.waitOn());
  const persistedApp = yield call([localStorage, localStorage.getItem], LOCAL_STORAGE_APP_KEY);
  if (persistedApp) {
    const dataApp = JSON.parse(persistedApp as string);
    yield put(actions.restoreGame(dataApp));
  }
  yield put(appActions.waitOff());
}

function* watchPersist(): Generator {
  const app = yield select(({ app }) => app);
  yield call([localStorage, localStorage.setItem], LOCAL_STORAGE_APP_KEY, JSON.stringify(app));
}

export function* appSaga() {
  yield takeEvery(actions.rehydrate.type, watchRehydrate);
  yield takeEvery(actions.setUserLevel.type, watchPersist);
  yield takeEvery(actions.moveToCell.type, watchPersist);
  yield takeEvery(actions.resetApp.type, watchPersist);
}
