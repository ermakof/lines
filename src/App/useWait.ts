import { actions } from '@src/App/appSlice';
import { useDispatch } from 'react-redux';

const useWait = () => {
  const dispatch = useDispatch();

  return (flag: boolean = true) => {
    dispatch(flag ? actions.waitOn() : actions.waitOff());
  };
};

export default useWait;
