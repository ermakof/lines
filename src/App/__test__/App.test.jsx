import React from 'react';
import { Provider, useDispatch, useSelector } from 'react-redux';
import { render, screen } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';

import App from '@src/App';
import initialState from '@src/App/initialState';

const mockStore = configureMockStore();

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

test('render App', () => {
  const store = mockStore(initialState);
  const mockDispatch = jest.fn();
  useDispatch.mockReturnValue(mockDispatch);
  useSelector.mockReturnValue(initialState);
  const { asFragment } = render(
    <Provider store={store}>
      <App />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
  const topPanel = screen.getByRole(/topPanel/gi);
  expect(topPanel).toBeInTheDocument();
  const gamePanel = screen.getByRole(/gamePanel/gi);
  expect(gamePanel).toBeInTheDocument();
  const bottomPanel = screen.getByRole(/bottomPanel/gi);
  expect(bottomPanel).toBeInTheDocument();
});
