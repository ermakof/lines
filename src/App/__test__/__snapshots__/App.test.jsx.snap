// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render App 1`] = `
<DocumentFragment>
  <div
    class="css-5simcc"
    role="rootApp"
  >
    <div
      class="css-1uv7a9d"
      role="topPanel"
    >
      <div
        class="css-ov1ktg"
      >
        <p
          class="css-z1cqd0"
        >
          Lines
        </p>
        <div
          class="css-qkmzfx"
          role="authForm"
        >
          <form
            action="#"
          >
            <label
              for="login"
            >
              Имя
            </label>
            <input
              id="login"
              name="login"
              placeholder="Введите логин"
              role="login"
              value=""
            />
            <label
              for="password"
            >
              Пароль
            </label>
            <input
              id="password"
              name="password"
              placeholder="Введите пароль"
              role="password"
              value=""
            />
            <button
              disabled=""
              role="buttonLogin"
              type="submit"
            >
              Войти
            </button>
          </form>
        </div>
      </div>
    </div>
    <div
      class="css-xt4orx"
      role="gamePanel"
    >
      <section
        class="css-1werspr"
        role="cellsGrid"
      >
        <div
          class="css-es8rv7"
          role="cellContainer-0"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-1"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-2"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-3"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-4"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-5"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-6"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-7"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-8"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-9"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-10"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-11"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-12"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-13"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-14"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-15"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-16"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-17"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-18"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-19"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-20"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-21"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-22"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-23"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-24"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-25"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-26"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-27"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-28"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-29"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-30"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-31"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-32"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-33"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-34"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-35"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-36"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-37"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-38"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-39"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-40"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-41"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-42"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-43"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-44"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-45"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-46"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-47"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-48"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-49"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-50"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-51"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-52"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-53"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-54"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-55"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-56"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-57"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-58"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-59"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-60"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-61"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-62"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-63"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-64"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-65"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-66"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-67"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-68"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-69"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-70"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-71"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-72"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-73"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-74"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-75"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-76"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-77"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-78"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-79"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-80"
        />
        <div
          style="clear: both;"
        />
      </section>
    </div>
    <div
      class="css-1563bnz"
    >
      <div
        class="css-1uv7a9d"
        role="bottomPanel"
      >
        <p
          class="css-o9gtt2"
          role="messageResult"
        >
          Результат:
        </p>
        <p
          class="css-o9gtt2"
          role="messagePercentFilled"
        >
          Процент заполнения: 0
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
